File: _scripts/preprocess-abilities.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Pre-build script to merge ability cache data into keybinds.yml
# This embeds the cached data directly into the keybinds file
# Run this before building Jekyll: ruby _scripts/preprocess-abilities.rb

require 'yaml'

KEYBINDS_FILE = '_data/keybinds.yml'
CACHE_FILE = '_data/abilities_cache.yml'
OUTPUT_FILE = '_data/keybinds.yml'

def load_yaml(file)
  YAML.load_file(file)
rescue => e
  puts "Error loading #{file}: #{e.message}"
  {}
end

def save_yaml(data, file)
  File.write(file, data.to_yaml)
end

def merge_cache_into_keybinds(keybinds, cache)
  return keybinds unless keybinds['classes'] && cache
  
  keybinds['classes'].each do |_class_key, class_data|
    next unless class_data['abilities']
    
    class_data['abilities'].each do |_key_id, ability|
      next unless ability.is_a?(Hash) && ability['wowhead_id']
      
      spell_id = ability['wowhead_id'].to_s
      # Try both string and integer keys
      cached = cache[spell_id] || cache[spell_id.to_i]
      
      if cached
        # Merge cached data into ability
        ability['icon'] = cached['icon']
        ability['tooltip'] = cached['tooltip']
        ability['cached_name'] = cached['name']
      end
    end
  end
  
  keybinds
end

def main
  puts "Preprocessing abilities..."
  puts "Loading keybinds from #{KEYBINDS_FILE}..."
  keybinds = load_yaml(KEYBINDS_FILE)
  
  puts "Loading cache from #{CACHE_FILE}..."
  cache = load_yaml(CACHE_FILE)
  
  if cache.empty?
    puts "Warning: Cache is empty. Run fetch-abilities-blizzard.rb first."
  else
    puts "Found #{cache.keys.length} cached abilities"
  end
  
  puts "Merging cache into keybinds..."
  merged = merge_cache_into_keybinds(keybinds, cache)
  
  puts "Saving to #{OUTPUT_FILE}..."
  save_yaml(merged, OUTPUT_FILE)
  
  puts "Done! Ability data is now embedded in keybinds.yml"
  puts "You can now build Jekyll and the icons will appear."
end

main if __FILE__ == $PROGRAM_NAME

File: scripts/preprocess-abilities.ts
#!/usr/bin/env node
// Ports _scripts/preprocess-abilities.rb — merges src/data/abilities_cache.yml
// into src/data/keybinds.yml (icon/tooltip/cached_name).
// Usage: node scripts/preprocess-abilities.ts

import fs from 'node:fs';
import { parseKeybinds, parseCache, serializeKeybinds } from '../src/lib/keybinds-io.ts';
import { KEYBINDS_PATH, CACHE_PATH } from './paths.ts';

function main() {
	console.log('Preprocessing abilities...');
	console.log(`Loading keybinds from ${KEYBINDS_PATH}...`);
	const keybinds = parseKeybinds(fs.readFileSync(KEYBINDS_PATH, 'utf8'));

	console.log(`Loading cache from ${CACHE_PATH}...`);
	const cache = parseCache(fs.readFileSync(CACHE_PATH, 'utf8'));

	const cacheSize = Object.keys(cache).length;
	if (cacheSize === 0) {
		console.log('Warning: Cache is empty. Run fetch-abilities.ts first.');
	} else {
		console.log(`Found ${cacheSize} cached abilities`);
	}

	console.log('Merging cache into keybinds...');
	for (const classData of Object.values(keybinds.classes)) {
		for (const ability of Object.values(classData.abilities ?? {})) {
			if (!ability?.wowhead_id) continue;
			const cached = cache[String(ability.wowhead_id)];
			if (cached) {
				ability.icon = cached.icon ?? undefined;
				ability.tooltip = cached.tooltip;
				ability.cached_name = cached.name;
			}
		}
	}

	console.log(`Saving to ${KEYBINDS_PATH}...`);
	fs.writeFileSync(KEYBINDS_PATH, serializeKeybinds(keybinds), 'utf8');

	console.log('Done! Ability data is now embedded in keybinds.yml');
	console.log('You can now build Astro and the icons will appear.');
}

main();
